include/seqan3/alignment/pairwise_alignment_edit_distance_unbanded.hpp

```diff
@@ -0,0 +1,548 @@
+    : public alignment_score_matrix<std::vector<int>>
+{
+    using alignment_t = pairwise_alignment_edit_distance_unbanded<database_t, query_t, align_config_t>;
+    using base_score_matrix_t = alignment_score_matrix<std::vector<int>>;
+    using score_t = int;
+
+    alignment_score_matrix(alignment_t const & alignment) :
+        base_score_matrix_t
+        {
+            [&]{
+                using word_t = typename alignment_t::word_type;
+                constexpr size_t word_size = sizeof(word_t)*8;
+
+                std::size_t _cols = alignment.database.size() + 1;
+                std::size_t _rows = alignment.query.size() + 1;
+                std::vector<int> scores{};
+                scores.reserve(_cols * _rows);
+
+                // init first row with 0, 1, 2, 3, ...
+                for(unsigned col=0; col < _cols; ++col)
+                    scores[col] = alignment_t::is_global ? col : 0;
+
+                auto deltas = [&](unsigned col)
+                {
+                    return [state = alignment.states[col]](unsigned row)
+                    {
+                        using bitset = std::bitset<word_size>;
+
+                        size_t chunk = row / word_size;
+                        size_t row_in_chunk = row % word_size;
+                        auto vp = state.vp[chunk];
+                        auto vn = state.vn[chunk];
+
+                        int p = bitset(vp)[row_in_chunk] ? 1 : 0;
+                        int n = bitset(vn)[row_in_chunk] ? 1 : 0;
+                        return p - n;
+                    };
+                };
+
+                for(size_t col=0; col < _cols; ++col)
+                {
+                    auto delta = deltas(col);
+                    for(size_t row=1; row < _rows; ++row)
+                        scores[row * _cols + col] = scores[(row-1) * _cols + col] + delta(row-1);
+                }
+
+                return scores;
+            }(),
+            alignment.query.size()+1,
+            alignment.database.size()+1
+        }
+    {
+    }
+};
+
+template<typename database_t, typename query_t, typename align_config_t>
+struct alignment_trace_matrix<pairwise_alignment_edit_distance_unbanded<database_t, query_t, align_config_t>>
+    : public alignment_trace_matrix<database_t const &, query_t const &, align_config_t, alignment_score_matrix<pairwise_alignment_edit_distance_unbanded<database_t, query_t, align_config_t>>>
+{
+    using alignment_t = pairwise_alignment_edit_distance_unbanded<database_t, query_t, align_config_t>;
+    using score_matrix_t = alignment_score_matrix<alignment_t>;
+    using base_trace_matrix_t = alignment_trace_matrix<database_t const &, query_t const &, align_config_t, score_matrix_t>;
+
+    alignment_trace_matrix(alignment_t const & alignment) :
+        base_trace_matrix_t{alignment.database, alignment.query, alignment.config, score_matrix_t{alignment}}
+    {
+    }
+};
```

Comment:
why do we need these many different matrix types?

Reply:
There are only two matrix types and they are needed. The specialisations are needed, because of the differences in the stored data between the algorithms.